### PR TITLE
find remote cluster ID for the filesystem

### PR DESCRIPTION
**Pull request by nitishkumar4**
## Pull request checklist

```release-notes

```

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

- User needs to pass clusterID along with the storage class.

## What is the new behavior?

- User do not need to pass clusterID along with the storage class.

## How risky is this change?
- [x] Small, isolated change
- [ ] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

**Comment by nitishkumar4** [comment]
@madhuthorat @deeghuge @amdabhad Please review.

File: driver/csiplugin/controllerserver.go [hunk 1]
```diff
@@ -591,20 +592,28 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
 		scaleVol.LocalFS = scaleVol.VolBackendFs
 	}
 
+	var remoteClusterID string
+	if volFsInfo.Type == filesystemTypeRemote {
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:596]: I think we can add the additional condition here something like 
if scaleVol.ClusterId == "" && volFsInfo.Type == filesystemTypeRemote

**Review by deeghuge — DISMISSED** [review]

File: driver/csiplugin/controllerserver.go (hunk 1 above)
Comment (nitishkumar4) [reply driver/csiplugin/controllerserver.go]: Added.

File: driver/csiplugin/controllerserver.go [hunk 2]
```diff
@@ -1818,3 +1827,108 @@ func (cs *ScaleControllerServer) ControllerExpandVolume(ctx context.Context, req
...
+					glog.V(5).Infof("updating cluster details in cache map for cluster %s.", scaleconfig.Clusters[i].ID)
+					cs.Driver.clusterMap.Store(ClusterName{cName}, ClusterDetails{cID, cName, time.Now(), 24})
+					cs.Driver.clusterMap.Store(ClusterID{cID}, ClusterDetails{cID, cName, time.Now(), 24})
+					if cName == clusterName {
+						return cID, nil
+					}
+				}
+			}
+
+			if !found {
```
Comment (madhuthorat) [inline driver/csiplugin/controllerserver.go:1914]: This is else part of "if found {" above, why not use else here?

File: driver/csiplugin/controllerserver.go [hunk 3]
```diff
@@ -1818,3 +1827,108 @@ func (cs *ScaleControllerServer) ControllerExpandVolume(ctx context.Context, req
...
+			if err != nil {
+				glog.V(5).Infof("unable to create new connector for the cluster %s", scaleconfig.Clusters[i].ID)
+				continue
+			}
+
+			cID = scaleconfig.Clusters[i].ID
+			glog.V(5).Infof("fetching cluster details from cache map for cluster %s", scaleconfig.Clusters[i].ID)
+			clusterDetails, found := cs.Driver.clusterMap.Load(ClusterID{cID})
+			if found {
+				glog.V(5).Infof("checking if cluster details found from cache map for cluster %s has expired.", scaleconfig.Clusters[i].ID)
```
Comment (madhuthorat) [inline driver/csiplugin/controllerserver.go:1887]: The lines of code starting here for "if found" part look similar/same as code above starting at line 1837. Can we have a function for these common lines of code ? Future maintenance of code becomes easy for any programmer.

**Review by madhuthorat — APPROVED** [review]
LGTM.
Please check couple of comments.

**Review by Jainbrt — APPROVED** [review]
LGTM.

File: driver/csiplugin/controllerserver.go (hunk 3 above)
Comment (nitishkumar4) [reply driver/csiplugin/controllerserver.go:1908]: created function `checkExpiry()` and `updateClusterMap()` to handle this. please review.

File: driver/csiplugin/controllerserver.go (hunk 2 above)
Comment (nitishkumar4) [reply driver/csiplugin/controllerserver.go]: using `else` now.

File: driver/csiplugin/controllerserver.go (hunk 3 above)
Comment (nitishkumar4) [reply driver/csiplugin/controllerserver.go:1908]: created function `checkExpiry()` and `updateClusterMap()` to handle this. please review.

**Comment by nitishkumar4** [comment]
regression test cases are passed. 
Command:
```
kubectl exec -it ibm-spectrum-scale-csi-test-pod -- pytest  remote_test.py -m "regression" --html=/data/regression_driver.html
```
Output:
```
12 passed, 337 deselected in 1240.96s (0:20:40) 
```

**Review by deeghuge — APPROVED** [review]

**Review by madhuthorat — APPROVED** [review]

**Pull request merged**